src/OpenTelemetry.Api/Trace/ActivityExtensions.cs

```diff
@@ -92,22 +87,30 @@ public static Status GetStatus(this Activity activity)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static object GetTagValue(this Activity activity, string tagName)
         {
-            var tagObjects = activity.TagObjects;
-            if (ReferenceEquals(tagObjects, EmptyActivityTagObjects))
-            {
-                return null;
-            }
+            Debug.Assert(activity != null, "Activity should not be null");
```

Comment:
Minor: consider the stock `Assert.IsNotNull`.

Reply:
You want me to add Xunit to API? 🤣 I don't think there's a similar method on System.Diagnostics.Debug, but there should be!

Reply:
Not familiar with .NET, I thought Assert.IsNotNull is a common thing and am surprised that it is not (so it looks like a compile time trick).

Reply:
Sorry didn't mean to poke fun 😄 Ya it's common to unit test frameworks but I don't think there's anything in the BCL.

Reply:
I took a look at the [doc](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assert?view=netcore-3.1) and it makes me wonder if we should use it or throw normal exception.

It seems in C++ the general trend is to avoid https://en.cppreference.com/w/c/error/assert from a library.

Reply:
.NET seems to be using it extensively for internal methods https://github.com/dotnet/runtime/blob/503d5602b01c0e362f1636d55da91ff3b8d909d8/src/libraries/System.Private.Xml/src/System/Xml/Core/IncrementalReadDecoders.cs#L61.

Reply:
So I guess:
1. use `Debug.Assert` for internal/private stuff, which benefits the developer working on the library.
2. use exception on public stuff as a contract for the library consumers, as they don't get a debug version of package.

Reply:
I can switch them to regular null checks if you would like. My thinking with the Debug.Assert was... user will have to work really hard to call these extensions with a null Activity. You have to go out of your way and do `OpenTelemetry.Trace.ActivityExtensions.GetTagValue(null, "myTag")`. If you are set on shooting yourself in the foot, who are we to stop you? 🤣 But for the 99.99% of callers who are doing `activity.GetTagValue("myTag")` it really will never be null, so I was trying to save the perf of doing an unnecessary check.

Reply:
I was trying to get your perspective and see if I need to change this https://github.com/open-telemetry/opentelemetry-dotnet/blob/4993ec61edceadca68d9f5b5387b59c7f85dcd63/src/OpenTelemetry/Trace/TracerProviderExtensions.cs#L25.

Based on the discussion, I think we should remove the null check for extension methods?

Reply:
It is kind of a risk/reward thing. That particular extension would only be called a handful of times (?) so it's probably fine to leave the check in there. If we don't have the check, static analysis tools like FxCop get upset so then we end up with suppressions which are kind of a maintainability headache. IMO we should only do this on hot-path stuff where we know there is a compelling benefit to justify the suppression. Assuming we get to turning FxCop on 😄 Poor @eddynaka has been trying to do that for a while.

Reply:
its getting harder and harder :(